odoo/addons/base/models/res_company.py

```diff
@@ -376,24 +376,74 @@ def _get_main_company(self):
     def __accessible_branches(self):
         # Get branches of this company that the current user can use
         self.ensure_one()
-        children = self.search([('id', 'child_of', self.id)])
+
+        rslt = []
+        current = self.sudo()
         accessible = self.env.companies
-        return (children & accessible).ids
+        while current:
+            rslt.extend((current & accessible).ids)
+            current = current.child_ids
+
+        if not rslt and self.env.su:
+            # Accessible companies will always be the same for super user when called in a cron.
+            # Because of that, the intersection between them and self might be empty. The super user anyway always has
+            # access to all companies (as it bypasses the record rules), so we return the current company in this case.
+            return self.ids
+
+        return rslt
 
     def _accessible_branches(self):
         return self.browse(self.__accessible_branches())
 
     def _all_branches_selected(self):
-        """Return whether or all the branches of the main company are selected.
+        """Return whether or all the branches of the companies in self are selected.
 
         Is ``True`` if all the branches, and only those, are selected.
-        Can be used when some actions only make sense for a whole company regardless of the
+        Can be used when some actions only make sense for whole companies regardless of the
         branches.
         """
-        return (
-            len(self.env.companies.root_id) == 1
-            and self.env.companies == self.env['res.company'].sudo().search([('id', 'child_of', self.env.company.root_id.ids)])
-        )
+        return self == self.sudo().search([('id', 'child_of', self.root_id.ids)])
+
+    def  _get_branches_with_same_vat(self, accessible_only=False):
```

Comment:
I think this is very specify to `account_reports` and should stay there

Reply:
Why is this not all just a search on `vat`, possibly coupled with `_accessible_branches`?